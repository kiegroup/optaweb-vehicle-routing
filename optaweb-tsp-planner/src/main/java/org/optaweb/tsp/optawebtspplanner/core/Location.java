package org.optaweb.tsp.optawebtspplanner.core;

import java.util.Objects;

/**
 * A unique location significant to the user.
 */
public class Location {

    private final long id;
    private final LatLng latLng;

    public Location(long id, LatLng latLng) {
        Objects.requireNonNull(latLng);
        this.id = id;
        this.latLng = latLng;
    }

    public long getId() {
        return id;
    }

    public LatLng getLatLng() {
        return latLng;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Location that = (Location) o;
        return id == that.id &&
                latLng.equals(that.latLng);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, latLng);
    }

    @Override
    public String toString() {
        return "Location{" +
                "id=" + id +
                ", latLng=" + latLng +
                '}';
    }
}
