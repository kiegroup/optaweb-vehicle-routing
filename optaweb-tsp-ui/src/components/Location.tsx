/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';

export interface ILocationProps {
  id: number;
  removeDisabled: boolean;
  removeHandler: (id: number) => void;
  selectHandler: (e: any) => void; // FIXME: Event Type
}

const Location: React.SFC<ILocationProps> = ({
  id,
  removeDisabled,
  removeHandler,
  selectHandler
}: ILocationProps) => {
  return (
    <div
      key={id}
      className='ma2 flex bg-animate hover-bg-light-gray'
      onMouseEnter={() => selectHandler(id)}
      onMouseLeave={() => selectHandler(NaN)}
    >
      <span className='w-80 pa2'>{`Location ${id}`}</span>
      <button
        type='button'
        disabled={removeDisabled}
        className='w-20 pa2'
        onClick={() => removeHandler(id)}
      >
        x
      </button>
    </div>
  );
};

export default Location;
