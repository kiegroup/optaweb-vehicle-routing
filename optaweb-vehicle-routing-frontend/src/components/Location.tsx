/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, Grid, GridItem } from '@patternfly/react-core';
import { TimesIcon } from '@patternfly/react-icons';
import * as React from 'react';

export interface ILocationProps {
  id: number;
  removeDisabled: boolean;
  removeHandler: (id: number) => void;
  selectHandler: (id: number) => void;
}

const Location: React.SFC<ILocationProps> = ({
  id,
  removeDisabled,
  removeHandler,
  selectHandler,
}: ILocationProps) => {
  return (
    <Grid
      gutter="md"
      onMouseEnter={() => selectHandler(id)}
      onMouseLeave={() => selectHandler(NaN)}
    >
      <GridItem span={9}>
        {`Location ${id}`}
      </GridItem>
      <GridItem span={3}>
        <Button
          variant="link"
          isDisabled={removeDisabled}
          onClick={() => removeHandler(id)}
          type="button"
        >
          <TimesIcon />
        </Button>
      </GridItem>
    </Grid>
  );
};

export default Location;
