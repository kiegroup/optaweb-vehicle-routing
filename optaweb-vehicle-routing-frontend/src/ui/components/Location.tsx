/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, DataListCell, DataListItem, DataListItemRow, Tooltip } from '@patternfly/react-core';
import { TimesIcon } from '@patternfly/react-icons';
import * as React from 'react';

export interface LocationProps {
  id: number;
  description: string | null;
  removeDisabled: boolean;
  removeHandler: (id: number) => void;
  selectHandler: (id: number) => void;
}

const Location: React.FC<LocationProps> = ({
  id,
  description,
  removeDisabled,
  removeHandler,
  selectHandler,
}) => {
  const [clicked, setClicked] = React.useState(false);

  function shorten(text: string) {
    const first = text.replace(/,.*/, '').trim();
    const short = first.substring(0, Math.min(20, first.length)).trim();
    if (short.length < first.length) {
      return `${short}...`;
    }
    return short;
  }

  return (
    <DataListItem
      isExpanded={false}
      aria-labelledby={`location-${id}`}
      onMouseEnter={() => selectHandler(id)}
      onMouseLeave={() => selectHandler(NaN)}
    >
      <DataListItemRow>
        <DataListCell isFilled>
          {(description && (
            <Tooltip content={description}>
              <span id={`location-${id}`}>{shorten(description)}</span>
            </Tooltip>
          ))
          || <span id={`location-${id}`}>{`Location ${id}`}</span>}
        </DataListCell>
        <DataListCell isFilled={false}>
          <Button
            type="button"
            variant="link"
            isDisabled={removeDisabled || clicked}
            onClick={() => {
              setClicked(true);
              removeHandler(id);
            }}
          >
            <TimesIcon />
          </Button>
        </DataListCell>
      </DataListItemRow>
    </DataListItem>
  );
};

export default Location;
