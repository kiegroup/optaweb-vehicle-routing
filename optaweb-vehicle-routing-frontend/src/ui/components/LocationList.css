.pf-c-data-list__cell {
  --pf-c-data-list__cell-cell--PaddingTop: var(--pf-global--spacer--sm);
  --pf-c-data-list__cell--PaddingTop: var(--pf-global--spacer--md);
  --pf-c-data-list__cell--PaddingBottom: var(--pf-global--spacer--sm);
}
