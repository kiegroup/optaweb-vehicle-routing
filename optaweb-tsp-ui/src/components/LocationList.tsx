/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, Card, CardBody, CardHeader } from '@patternfly/react-core';
import * as React from 'react';
import { ITSPRoute } from '../store/tsp/types';
import Location from './Location';
import TripData from './TripData';

export interface ILocationListProps extends ITSPRoute {
  removeHandler: (id: number) => void;
  selectHandler: (e: any) => void;
  loadHandler: () => void;
  maxDistance: number;
}

const renderEmptyLocationList = ({ loadHandler }: ILocationListProps) => {
  return (
    <Card>
      <CardHeader>Click map to add locations</CardHeader>
      <CardBody>
        <Button type="button" style={{ width: '100%' }} onClick={loadHandler}>
          Load 40 European cities
        </Button>
      </CardBody>
    </Card>
  );
};

const renderLocationList = ({
  route,
  domicileId,
  distance = '',
  removeHandler,
  selectHandler,
  loadHandler,
  maxDistance,
}: ILocationListProps) => {
  return (
    <Card>
      <CardHeader>
        Distance: {distance}
        <br />
        Locations: {route.length}
        <TripData
          maxDistance={maxDistance}
          distance={parseInt(distance, 10) || maxDistance}
        />
        <hr />
      </CardHeader>
      <CardBody>
        {/*
               The calculated maxHeight is a hack because the constant 116px depends
               on the height of Distance and Locations rows (above) and individual location rows.
               */}
        <div style={{ maxHeight: 'calc(100vh - 195px)', overflowY: 'auto' }}>
          {route
            .slice(0) // clone the array because
            // sort is done in place (that would affect the route)
            .sort((a, b) => a.id - b.id)
            .map(location => (
              <Location
                key={location.id}
                id={location.id}
                removeDisabled={route.length > 1 && location.id === domicileId}
                removeHandler={removeHandler}
                selectHandler={selectHandler}
              />
            ))}
        </div>
      </CardBody>
    </Card>
  );
};

const LocationList: React.SFC<ILocationListProps> = (
  props: ILocationListProps
) => {
  return (
    <div
      className="leaflet-top leaflet-left leaflet-touch"
      style={{ zIndex: 500 }}
    >
      <div className="leaflet-control leaflet-bar">
        {props.route.length === 0
          ? renderEmptyLocationList(props)
          : renderLocationList(props)}
      </div>
    </div>
  );
};

export default LocationList;
