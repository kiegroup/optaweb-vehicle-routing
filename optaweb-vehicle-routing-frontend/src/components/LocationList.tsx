/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, Card, CardBody, CardHeader } from '@patternfly/react-core';
import * as React from 'react';
import { IRoute } from '../store/route/types';
import Location from './Location';
import SearchBox from './SearchBox';
import TripData from './TripData';

export interface ILocationListProps {
  removeHandler: (id: number) => void;
  selectHandler: (id: number) => void;
  loadHandler: () => void;
  clearHandler: () => void;
  maxDistance: number;
  route: IRoute;
  domicileId: number;
  isDemoLoading: boolean;
}

const renderEmptyLocationList = ({
  loadHandler,
  isDemoLoading,
}: ILocationListProps) => {
  return (
    <Card>
      <CardHeader>Click map to add locations</CardHeader>
      <CardBody>
        <Button
          type="button"
          isDisabled={isDemoLoading}
          style={{ width: '100%' }}
          onClick={loadHandler}
        >
          Load demo
        </Button>
      </CardBody>
    </Card>
  );
};

const renderLocationList = ({
  route: { distance, locations },
  domicileId,
  removeHandler,
  selectHandler,
  clearHandler,
  maxDistance,
  isDemoLoading,
}: ILocationListProps) => {
  return (
    <Card>
      <CardHeader>
        Distance: {distance}
        <br />
        Locations: {locations.length}
        <TripData
          maxDistance={maxDistance}
          distance={parseInt(distance, 10) || maxDistance}
        />
        <br />
        <Button
          type="button"
          isDisabled={isDemoLoading}
          style={{ width: '100%' }}
          onClick={clearHandler}
        >
          Clear
        </Button>
        <br />
        <SearchBox />
      </CardHeader>
      <CardBody>
        {/*
        The calculated maxHeight is a hack because the last constant depends
        on the height of CardHeader (above).
        */}
        <div style={{ maxHeight: 'calc(100vh - 24px - 24px - 8px - 196px)', overflowY: 'auto' }}>
          {locations
            .slice(0) // clone the array because
            // sort is done in place (that would affect the route)
            .sort((a, b) => a.id - b.id)
            .map(location => (
              <Location
                key={location.id}
                id={location.id}
                removeDisabled={locations.length > 1 && location.id === domicileId}
                removeHandler={removeHandler}
                selectHandler={selectHandler}
              />
            ))}
        </div>
      </CardBody>
    </Card>
  );
};

const LocationList: React.SFC<ILocationListProps> = (
  props: ILocationListProps,
) => {
  return (
    <div
      className="leaflet-top leaflet-left leaflet-touch"
      style={{ zIndex: 500 }}
    >
      <div className="leaflet-control leaflet-bar">
        {props.route.locations.length === 0
          ? renderEmptyLocationList(props)
          : renderLocationList(props)}
      </div>
    </div>
  );
};

export default LocationList;
