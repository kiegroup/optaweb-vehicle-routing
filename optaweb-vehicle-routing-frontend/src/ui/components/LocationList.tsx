/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Bullseye, DataList } from '@patternfly/react-core';
import * as React from 'react';
import { Location } from 'store/route/types';
import LocationItem from './Location';
import './LocationList.css';

export interface LocationListProps {
  removeHandler: (id: number) => void;
  selectHandler: (id: number) => void;
  depot: Location | null;
  visits: Location[];
}

const renderEmptyLocationList: React.FC<LocationListProps> = () => (
  <DataList aria-label="Empty location list">
    <Bullseye>No locations</Bullseye>
  </DataList>
);

const renderLocationList: React.FC<LocationListProps> = ({
  depot,
  visits,
  removeHandler,
  selectHandler,
}) => (
  <div style={{ overflowY: 'auto' }}>
    <DataList
      aria-label="List of locations"
    >
      {depot && (
        <LocationItem
          key={depot.id}
          id={depot.id}
          description={depot.description || null}
          removeDisabled={visits.length > 0}
          removeHandler={removeHandler}
          selectHandler={selectHandler}
        />
      )}
      {visits
        .slice(0) // clone the array because
        // sort is done in place (that would affect the route)
        .sort((a, b) => a.id - b.id)
        .map(visit => (
          <LocationItem
            key={visit.id}
            id={visit.id}
            description={visit.description || null}
            removeDisabled={false}
            removeHandler={removeHandler}
            selectHandler={selectHandler}
          />
        ))}
    </DataList>
  </div>
);

const LocationList: React.FC<LocationListProps> = props => (
  props.visits.length === 0 && props.depot === null
    ? renderEmptyLocationList(props)
    : renderLocationList(props)
);

export default LocationList;
