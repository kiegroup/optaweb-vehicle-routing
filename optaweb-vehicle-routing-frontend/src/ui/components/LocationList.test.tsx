/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import Location from './Location';
import LocationList, { LocationListProps } from './LocationList';

describe('Location List Component', () => {
  it('should render correctly with no routes', () => {
    const props: LocationListProps = {
      removeHandler: jest.fn(),
      selectHandler: jest.fn(),
      depot: null,
      visits: [],
    };
    const locationList = shallow(<LocationList {...props} />);
    expect(toJson(locationList)).toMatchSnapshot();
  });

  it('should render correctly with a few routes', () => {
    const props: LocationListProps = {
      removeHandler: jest.fn(),
      selectHandler: jest.fn(),
      depot: {
        id: 1,
        lat: 1.345678,
        lng: 1.345678,
        description: 'Depot',
      },
      visits: [
        {
          id: 2,
          lat: 2.345678,
          lng: 2.345678,
          description: 'Visit 1',
        },
        {
          id: 3,
          lat: 3.676111,
          lng: 3.568333,
          description: 'Visit 2',
        },
      ],
    };
    const locationList = shallow(<LocationList {...props} />);
    expect(toJson(locationList)).toMatchSnapshot();

    expect(locationList.find(Location)).toHaveLength(props.visits.length + 1);
  });
});
