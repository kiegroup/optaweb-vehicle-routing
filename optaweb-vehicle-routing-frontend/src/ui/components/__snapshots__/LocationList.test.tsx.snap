// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location List Component should render correctly with a few routes 1`] = `
<div
  data-cy="location-list"
  style={
    Object {
      "overflowY": "auto",
    }
  }
>
  <DataList
    aria-label="List of locations"
  >
    <Location
      description="Depot"
      id={1}
      key="1"
      removeDisabled={true}
      removeHandler={[MockFunction]}
      selectHandler={[MockFunction]}
    />
    <Location
      description="Visit 1"
      id={2}
      key="2"
      removeDisabled={false}
      removeHandler={[MockFunction]}
      selectHandler={[MockFunction]}
    />
    <Location
      description="Visit 2"
      id={3}
      key="3"
      removeDisabled={false}
      removeHandler={[MockFunction]}
      selectHandler={[MockFunction]}
    />
  </DataList>
</div>
`;

exports[`Location List Component should render correctly with no routes 1`] = `
<DataList
  aria-label="Empty location list"
>
  <Bullseye>
    No locations
  </Bullseye>
</DataList>
`;
